File: clients/go/internal/embed/potion/tokenizer.go
package potion

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"unicode/utf8"
)

// Tokenizer is the part of a tokenizers tokenizer.json that model2vec's
// encode_batch_fast(texts, add_special_tokens=False) runs for a potion
// model: added tokens split out of the raw text, BertNormalizer,
// BertPreTokenizer and WordPiece. Any other component is refused at load,
// never guessed at.
type Tokenizer struct {
	vocab     map[string]int
	unk       string
	unkID     int
	prefix    string
	maxChars  int
	added     []addedToken // longest first
	allTokens map[string]int
}

type addedToken struct {
	content string
	id      int
}

type rawTokenizer struct {
	Truncation    json.RawMessage `json:"truncation"`
	Padding       json.RawMessage `json:"padding"`
	AddedTokens   []rawAdded      `json:"added_tokens"`
	Normalizer    json.RawMessage `json:"normalizer"`
	PreTokenizer  json.RawMessage `json:"pre_tokenizer"`
	Model         json.RawMessage `json:"model"`
	PostProcessor json.RawMessage `json:"post_processor"`
	Decoder       json.RawMessage `json:"decoder"`
}

type rawAdded struct {
	ID         int    `json:"id"`
	Content    string `json:"content"`
	SingleWord bool   `json:"single_word"`
	LStrip     bool   `json:"lstrip"`
	RStrip     bool   `json:"rstrip"`
	Normalized bool   `json:"normalized"`
	Special    bool   `json:"special"`
}

func isNull(m json.RawMessage) bool { return len(m) == 0 || string(m) == "null" }

// ParseTokenizer reads tokenizer.json.
func ParseTokenizer(data []byte) (*Tokenizer, error) {
	var raw rawTokenizer
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("tokenizer.json: %v", err)
	}
	if !isNull(raw.Truncation) || !isNull(raw.Padding) {
		return nil, fmt.Errorf("tokenizer.json sets truncation or padding, which this binary does not model")
	}
	var norm struct {
		Type               string `json:"type"`
		CleanText          bool   `json:"clean_text"`
		HandleChineseChars bool   `json:"handle_chinese_chars"`
		StripAccents       *bool  `json:"strip_accents"`
		Lowercase          bool   `json:"lowercase"`
	}
	if err := json.Unmarshal(raw.Normalizer, &norm); err != nil || norm.Type != "BertNormalizer" ||
		!norm.CleanText || !norm.HandleChineseChars || !norm.Lowercase ||
		(norm.StripAccents != nil && !*norm.StripAccents) {
		return nil, fmt.Errorf("tokenizer.json has a normalizer this binary does not model (only BertNormalizer with clean_text, handle_chinese_chars and lowercase)")
	}
	var pre struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(raw.PreTokenizer, &pre); err != nil || pre.Type != "BertPreTokenizer" {
		return nil, fmt.Errorf("tokenizer.json has a pre_tokenizer this binary does not model (only BertPreTokenizer)")
	}
	var model struct {
		Type     string         `json:"type"`
		Unk      *string        `json:"unk_token"`
		Prefix   string         `json:"continuing_subword_prefix"`
		MaxChars int            `json:"max_input_chars_per_word"`
		Vocab    map[string]int `json:"vocab"`
	}
	if err := json.Unmarshal(raw.Model, &model); err != nil || model.Type != "WordPiece" || model.Unk == nil {
		return nil, fmt.Errorf("tokenizer.json has a model this binary does not model (only WordPiece with an unk_token)")
	}
	unkID, ok := model.Vocab[*model.Unk]
	if !ok {
		return nil, fmt.Errorf("tokenizer.json: the unk_token %q is not in the vocabulary", *model.Unk)
	}
	t := &Tokenizer{
		vocab: model.Vocab, unk: *model.Unk, unkID: unkID, prefix: model.Prefix, maxChars: model.MaxChars,
		allTokens: map[string]int{},
	}
	for k, v := range model.Vocab {
		t.allTokens[k] = v
	}
	for _, a := range raw.AddedTokens {
		if a.SingleWord || a.LStrip || a.RStrip || a.Normalized || a.Content == "" {
			return nil, fmt.Errorf("tokenizer.json has an added token (%q) with options this binary does not model", a.Content)
		}
		t.added = append(t.added, addedToken{a.Content, a.ID})
		t.allTokens[a.Content] = a.ID
	}
	sort.SliceStable(t.added, func(i, j int) bool { return len(t.added[i].content) > len(t.added[j].content) })
	return t, nil
}

// UnkID is the id of the unknown token, which model2vec drops.
func (t *Tokenizer) UnkID() int { return t.unkID }

// Vocab is get_vocab(): the model's vocabulary with the added tokens.
func (t *Tokenizer) Vocab() map[string]int { return t.allTokens }

// Encode is encode(text, add_special_tokens=False).ids.
func (t *Tokenizer) Encode(text string) []int {
	var ids []int
	for len(text) > 0 {
		at, tok := t.findAdded(text)
		if at < 0 {
			ids = t.encodeSegment(text, ids)
			break
		}
		ids = t.encodeSegment(text[:at], ids)
		ids = append(ids, tok.id)
		text = text[at+len(tok.content):]
	}
	return ids
}

// findAdded is the leftmost added token in text, the longest one at that
// place.
func (t *Tokenizer) findAdded(text string) (int, addedToken) {
	best, bestTok := -1, addedToken{}
	for _, a := range t.added {
		i := strings.Index(text, a.content)
		if i >= 0 && (best < 0 || i < best) {
			best, bestTok = i, a
		}
	}
	return best, bestTok
}

func (t *Tokenizer) encodeSegment(seg string, ids []int) []int {
	if seg == "" {
		return ids
	}
	for _, word := range preTokenize(normalize(seg)) {
		ids = t.wordPiece(word, ids)
	}
	return ids
}

// wordPiece is WordPiece.tokenize for one word: greedy longest match,
// the prefix on every piece after the first, the whole word unknown when
// one piece has no match or the word is too long.
func (t *Tokenizer) wordPiece(word string, ids []int) []int {
	if utf8.RuneCountInString(word) > t.maxChars {
		return append(ids, t.unkID)
	}
	var sub []int
	start := 0
	for start < len(word) {
		end := len(word)
		found := -1
		for start < end {
			s := word[start:end]
			if start > 0 {
				s = t.prefix + s
			}
			if id, ok := t.vocab[s]; ok {
				found = id
				break
			}
			_, size := utf8.DecodeLastRuneInString(word[start:end])
			end -= size
		}
		if found < 0 {
			return append(ids, t.unkID)
		}
		sub = append(sub, found)
		start = end
	}
	return append(ids, sub...)
}

type rng struct{ lo, hi rune }

type mapped struct {
	r rune
	s string
}

func lookupMapped(r rune) (string, bool) {
	i := sort.Search(len(normMapped), func(i int) bool { return normMapped[i].r >= r })
	if i < len(normMapped) && normMapped[i].r == r {
		return normMapped[i].s, true
	}
	return "", false
}

func inTable(tab []rng, r rune) bool {
	i := sort.Search(len(tab), func(i int) bool { return tab[i].hi >= r })
	return i < len(tab) && tab[i].lo <= r
}

// normalize is BertNormalizer, one code point at a time.
func normalize(s string) string {
	var b strings.Builder
	for _, r := range s {
		switch {
		case inTable(normRemoved, r):
		case inTable(normSpace, r):
			b.WriteByte(' ')
		case inTable(normCJK, r):
			b.WriteByte(' ')
			b.WriteRune(r)
			b.WriteByte(' ')
		case r >= 0xAC00 && r <= 0xD7A3:
			n := r - 0xAC00
			b.WriteRune(0x1100 + n/588)
			b.WriteRune(0x1161 + (n%588)/28)
			if tail := 0x11A7 + n%28; tail != 0x11A7 {
				b.WriteRune(tail)
			}
		default:
			if m, ok := lookupMapped(r); ok {
				b.WriteString(m)
			} else {
				b.WriteRune(r)
			}
		}
	}
	return b.String()
}

// preTokenize is BertPreTokenizer: split on whitespace (dropped), then
// every punctuation mark is a word of its own.
func preTokenize(s string) []string {
	var out []string
	start := -1
	for i, r := range s {
		switch {
		case inTable(preSpace, r):
			if start >= 0 {
				out = append(out, s[start:i])
				start = -1
			}
		case inTable(prePunct, r):
			if start >= 0 {
				out = append(out, s[start:i])
				start = -1
			}
			out = append(out, string(r))
		default:
			if start < 0 {
				start = i
			}
		}
	}
	if start >= 0 {
		out = append(out, s[start:])
	}
	return out
}

File: clients/rust/src/pathways/potion/tokenizer.rs
//! The part of a tokenizers `tokenizer.json` that model2vec's
//! `encode_batch_fast(texts, add_special_tokens=False)` runs for a potion
//! model: added tokens split out of the raw text, BertNormalizer,
//! BertPreTokenizer and WordPiece. Any other component is refused at
//! load, never guessed at.

use std::collections::HashMap;

use super::tables::{NORM_CJK, NORM_MAPPED, NORM_REMOVED, NORM_SPACE, PRE_PUNCT, PRE_SPACE};

pub struct Tokenizer {
    vocab: HashMap<String, usize>,
    pub unk_id: usize,
    prefix: String,
    max_chars: usize,
    /// Longest first.
    added: Vec<(String, usize)>,
    all: HashMap<String, usize>,
}

fn is_null(v: Option<&serde_json::Value>) -> bool {
    matches!(v, None | Some(serde_json::Value::Null))
}

fn id_of(v: &serde_json::Value) -> Option<usize> {
    v.as_u64().map(|n| n as usize)
}

impl Tokenizer {
    /// Reads tokenizer.json.
    pub fn parse(data: &[u8]) -> Result<Tokenizer, String> {
        let raw: serde_json::Value = serde_json::from_slice(data).map_err(|e| format!("tokenizer.json: {e}"))?;
        let get = |k: &str| raw.get(k);
        if !is_null(get("truncation")) || !is_null(get("padding")) {
            return Err("tokenizer.json sets truncation or padding, which this binary does not model".into());
        }
        let norm = get("normalizer").cloned().unwrap_or(serde_json::Value::Null);
        let flag = |k: &str| norm.get(k).and_then(|v| v.as_bool()).unwrap_or(false);
        let strip = norm.get("strip_accents");
        let strip_ok = matches!(strip, None | Some(serde_json::Value::Null) | Some(serde_json::Value::Bool(true)));
        if norm.get("type").and_then(|t| t.as_str()) != Some("BertNormalizer")
            || !flag("clean_text")
            || !flag("handle_chinese_chars")
            || !flag("lowercase")
            || !strip_ok
        {
            return Err("tokenizer.json has a normalizer this binary does not model (only BertNormalizer with \
                        clean_text, handle_chinese_chars and lowercase)"
                .into());
        }
        let pre = get("pre_tokenizer").and_then(|p| p.get("type")).and_then(|t| t.as_str());
        if pre != Some("BertPreTokenizer") {
            return Err("tokenizer.json has a pre_tokenizer this binary does not model (only BertPreTokenizer)".into());
        }
        let model = get("model").cloned().unwrap_or(serde_json::Value::Null);
        let unk = model.get("unk_token").and_then(|u| u.as_str());
        let vocab_raw = model.get("vocab").and_then(|v| v.as_object());
        let (unk, vocab_raw) = match (model.get("type").and_then(|t| t.as_str()), unk, vocab_raw) {
            (Some("WordPiece"), Some(u), Some(v)) => (u.to_string(), v),
            _ => {
                return Err("tokenizer.json has a model this binary does not model (only WordPiece with an unk_token)".into())
            }
        };
        let mut vocab = HashMap::new();
        for (k, v) in vocab_raw {
            let id = id_of(v).ok_or("tokenizer.json: a vocabulary id is not a number")?;
            vocab.insert(k.clone(), id);
        }
        let unk_id = *vocab.get(&unk).ok_or_else(|| format!("tokenizer.json: the unk_token {unk:?} is not in the vocabulary"))?;
        let prefix = model.get("continuing_subword_prefix").and_then(|p| p.as_str()).unwrap_or("").to_string();
        let max_chars = model.get("max_input_chars_per_word").and_then(|p| p.as_u64()).unwrap_or(0) as usize;
        let mut all = vocab.clone();
        let mut added = vec![];
        if let Some(list) = get("added_tokens").and_then(|a| a.as_array()) {
            for a in list {
                let b = |k: &str| a.get(k).and_then(|v| v.as_bool()).unwrap_or(false);
                let content = a.get("content").and_then(|c| c.as_str()).unwrap_or("").to_string();
                if b("single_word") || b("lstrip") || b("rstrip") || b("normalized") || content.is_empty() {
                    return Err(format!("tokenizer.json has an added token ({content:?}) with options this binary does not model"));
                }
                let id = a.get("id").and_then(id_of).ok_or("tokenizer.json: an added token has no id")?;
                all.insert(content.clone(), id);
                added.push((content, id));
            }
        }
        added.sort_by(|a, b| b.0.len().cmp(&a.0.len()));
        Ok(Tokenizer { vocab, unk_id, prefix, max_chars, added, all })
    }

    /// `get_vocab()`: the model's vocabulary with the added tokens.
    pub fn vocab(&self) -> &HashMap<String, usize> {
        &self.all
    }

    /// `encode(text, add_special_tokens=False).ids`.
    pub fn encode(&self, text: &str) -> Vec<usize> {
        let mut ids = vec![];
        let mut text = text;
        while !text.is_empty() {
            match self.find_added(text) {
                None => {
                    self.segment(text, &mut ids);
                    break;
                }
                Some((at, content, id)) => {
                    self.segment(&text[..at], &mut ids);
                    ids.push(id);
                    text = &text[at + content.len()..];
                }
            }
        }
        ids
    }

    /// The leftmost added token in `text`, the longest one at that place.
    fn find_added<'a>(&'a self, text: &str) -> Option<(usize, &'a str, usize)> {
        let mut best: Option<(usize, &str, usize)> = None;
        for (c, id) in &self.added {
            if let Some(i) = text.find(c.as_str()) {
                if best.is_none_or(|b| i < b.0) {
                    best = Some((i, c, *id));
                }
            }
        }
        best
    }

    fn segment(&self, seg: &str, ids: &mut Vec<usize>) {
        if seg.is_empty() {
            return;
        }
        for word in pre_tokenize(&normalize(seg)) {
            self.word_piece(word, ids);
        }
    }

    /// WordPiece.tokenize for one word: greedy longest match, the prefix
    /// on every piece after the first, the whole word unknown when one
    /// piece has no match or the word is too long.
    fn word_piece(&self, word: &str, ids: &mut Vec<usize>) {
        if word.chars().count() > self.max_chars {
            ids.push(self.unk_id);
            return;
        }
        let mut sub = vec![];
        let mut start = 0;
        let mut buf = String::new();
        while start < word.len() {
            let mut end = word.len();
            let mut found = None;
            while start < end {
                let piece = &word[start..end];
                let key: &str = if start > 0 {
                    buf.clear();
                    buf.push_str(&self.prefix);
                    buf.push_str(piece);
                    &buf
                } else {
                    piece
                };
                if let Some(&id) = self.vocab.get(key) {
                    found = Some(id);
                    break;
                }
                let last = word[start..end].chars().next_back().map_or(1, |c| c.len_utf8());
                end -= last;
            }
            match found {
                None => {
                    ids.push(self.unk_id);
                    return;
                }
                Some(id) => {
                    sub.push(id);
                    start = end;
                }
            }
        }
        ids.extend(sub);
    }
}

fn in_table(tab: &[(u32, u32)], c: u32) -> bool {
    let i = tab.partition_point(|&(_, hi)| hi < c);
    i < tab.len() && tab[i].0 <= c
}

fn mapped(c: u32) -> Option<&'static str> {
    NORM_MAPPED.binary_search_by_key(&c, |&(k, _)| k).ok().map(|i| NORM_MAPPED[i].1)
}

/// BertNormalizer, one code point at a time.
pub fn normalize(s: &str) -> String {
    let mut b = String::with_capacity(s.len());
    for ch in s.chars() {
        let c = ch as u32;
        if in_table(NORM_REMOVED, c) {
        } else if in_table(NORM_SPACE, c) {
            b.push(' ');
        } else if in_table(NORM_CJK, c) {
            b.push(' ');
            b.push(ch);
            b.push(' ');
        } else if (0xAC00..=0xD7A3).contains(&c) {
            let n = c - 0xAC00;
            b.push(char::from_u32(0x1100 + n / 588).unwrap_or(ch));
            b.push(char::from_u32(0x1161 + (n % 588) / 28).unwrap_or(ch));
            let tail = 0x11A7 + n % 28;
            if tail != 0x11A7 {
                b.push(char::from_u32(tail).unwrap_or(ch));
            }
        } else if let Some(m) = mapped(c) {
            b.push_str(m);
        } else {
            b.push(ch);
        }
    }
    b
}

/// BertPreTokenizer: split on white space (dropped), then every
/// punctuation mark is a word of its own.
pub fn pre_tokenize(s: &str) -> Vec<&str> {
    let mut out = vec![];
    let mut start: Option<usize> = None;
    for (i, ch) in s.char_indices() {
        let c = ch as u32;
        if in_table(PRE_SPACE, c) {
            if let Some(st) = start.take() {
                out.push(&s[st..i]);
            }
        } else if in_table(PRE_PUNCT, c) {
            if let Some(st) = start.take() {
                out.push(&s[st..i]);
            }
            out.push(&s[i..i + ch.len_utf8()]);
        } else if start.is_none() {
            start = Some(i);
        }
    }
    if let Some(st) = start {
        out.push(&s[st..]);
    }
    out
}
